File: backend/app/schemas/progress.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class Achievement(BaseModel):
    id: str
    label: str
    description: str
    unlocked: bool
    progress_current: float
    progress_target: float


class RecentActivityItem(BaseModel):
    letter: str
    status: str  # pass, fail, no_attempt_detected
    confidence: Optional[float] = None
    created_at: Optional[datetime] = None


class PerformanceForecast(BaseModel):
    available: bool
    # Set only when available is False, explaining what's missing.
    reason: Optional[str] = None
    current_level: Optional[str] = None
    predicted_next_level: Optional[str] = None
    trend_percent_per_day: Optional[float] = None
    estimated_days_to_next_level: Optional[int] = None


class LearnerProgressResponse(BaseModel):
    learner_id: str
    current_streak_days: int
    achievements: list[Achievement]
    recent_activity: list[RecentActivityItem]
    forecast: PerformanceForecast

File: frontend/src/types/progress.ts
// Mirrors backend/app/schemas/progress.py

export interface Achievement {
  id: string;
  label: string;
  description: string;
  unlocked: boolean;
  progress_current: number;
  progress_target: number;
}

export interface RecentActivityItem {
  letter: string;
  status: 'pass' | 'fail' | 'no_attempt_detected';
  confidence: number | null;
  created_at: string | null;
}

export interface PerformanceForecast {
  available: boolean;
  // Set only when available is false, explaining what's missing — never
  // paired with fabricated numbers below it.
  reason: string | null;
  current_level: string | null;
  predicted_next_level: string | null;
  trend_percent_per_day: number | null;
  estimated_days_to_next_level: number | null;
}

export interface LearnerProgress {
  learner_id: string;
  current_streak_days: number;
  achievements: Achievement[];
  recent_activity: RecentActivityItem[];
  forecast: PerformanceForecast;
}
